### ISSUE TITLE
Research: multi-verifier compute sidecars for Needle Stage B

### ISSUE THREAD

**Issue by TeaShaman-cyber** [TeaShaman-cyber/theseus-needle-lab#33]
## Research question

Can Needle Stage B offload bounded verification/post-processing work to external mathematical sidecars without weakening the experiment's causal or provenance contract?

This Issue is downstream of #26 and does **not** authorize replacing Needle/JAX training or launching quality replicas by itself.

## Motivation

Stage B resource work exposed two distinct concerns:

1. GitHub-hosted CPU wall time varies materially across runner classes, so non-training work should not consume runner budget unnecessarily when a bounded external verifier can handle it.
2. Independent verification is valuable in its own right: a second compute path can certify acceptance conditions without becoming training authority.

The existing Theseus Public Observatory Wolfram MCP pilot provides a proven reference pattern: claim JSON -> external verifier -> normalized receipt -> claim graph, with verifier/transport failure kept separate from mathematical falsification.

### A. Wolfram MCP

Use the official Wolfram MCP / `WolframLanguageEvaluator` as a general symbolic/exact verifier for bounded mathematical claims.

Candidate Stage B uses:

- exact acceptance inequalities;
- confusion/count checks;
- R1/R2 acceptance-condition verification;
- independent resource/capacity calculations;
- bounded statistical post-processing;
- optional small gradient/update witness calculations where the required state can be represented compactly.

Wolfram is **not** proposed as the primary Needle/JAX training runtime.

### B. Precise Special Functions

Use the connected high-precision special-functions provider as a narrow independent numerical oracle for supported functions:

- Gamma;
- Riemann zeta;
- Bessel J/Y/I/K;
- Gauss hypergeometric 2F1;
- complete elliptic K/E.

This verifier is intentionally narrower than Wolfram. It is useful only when a Stage B/adjacent claim actually depends on one of these functions or when an independent high-precision numerical certificate is scientifically meaningful.

It must **not** be used merely to manufacture a second opinion for integer counts, byte contracts, or unrelated algebra.

### C. Repository-local deterministic verifier

Remain authoritative for:

- exact bytes and hashes;
- JSON/schema key order where behavioral bytes matter;
- dataset geometry and discrete identities;
- provenance bindings;
- train/held-out leakage barriers;
- execution identity/readback.

## Proposed architecture

```text
Needle/JAX training
      |
      +--> repo-local deterministic verifier
      |      bytes / hashes / manifests / discrete contracts
      |
      +--> Wolfram MCP sidecar
      |      symbolic/exact/statistical bounded claims
      |
      +--> Precise Special Functions sidecar
             narrow high-precision numerical certificates

all verifier outputs
      -> typed receipts
      -> explicit verifier identity/version/interface
      -> source/input digests where applicable
      -> fail-closed claim graph / final Stage B receipt
```

## Authority boundary

External verifiers are **witnesses, not training authority**.

A sidecar result must never silently mutate:

- model weights;
- training data;
- hyperparameters;
- experiment disposition rules.

Any disagreement between independent verifiers is preserved as first-class evidence and resolves to `INCONCLUSIVE_VERIFIER_DIVERGENCE` or a more specific bounded status, never by choosing the preferred answer after the fact.

## Initial experiments

1. Reproduce one existing Stage B discrete acceptance calculation locally and through Wolfram, comparing typed receipts.
2. Identify whether any current Stage B math actually benefits from the special-functions oracle. If no supported function is materially present, record `NOT_APPLICABLE` rather than forcing integration.
3. Probe a tiny bounded LoRA/JAX gradient or update witness only if its inputs can be represented without large tensor transfer; compare local result with an external mathematical formulation.
4. Measure runner work saved by externalizing post-processing separately from scientific verification value.

## Evidence contract

Each external-verifier receipt should bind at minimum:

```text
claim_id
verifier_kind
verifier_interface/version
input identity/digest
requested precision/constraints
raw returned value or durable response evidence where available
normalized result
comparison rule/tolerance if numerical
transport status
verification disposition
scope / does_not_establish
```

The reviewer lesson from #21 applies: a normalized result digest is not automatically proof of the exact source/response bytes used to derive it. Preserve raw evidence when the interface makes that possible; otherwise mark the limitation explicitly.

## Acceptance criteria

This research line succeeds if it establishes a small, reusable multi-verifier adapter contract that:

- keeps Needle/JAX training unchanged;
- can execute at least one bounded Wolfram Stage B claim with a reproducible receipt;
- either executes one scientifically relevant special-functions claim or records the provider as `NOT_APPLICABLE` for current Stage B;
- preserves local byte/provenance authority;
- reports verifier disagreement and transport failure distinctly;
- demonstrates whether any GitHub runner work is actually saved.

## Relationship to #26

#26 remains the authority for Stage B training/evaluation. This Issue may strengthen verification and reduce bounded post-processing load, but quality replicas remain gated by #26's corrected schema/projection and execution contracts.

---

Prepared collaboratively by Semyon Poklad & Jester (ChatGPT), Theseus research project.

**Comment by TeaShaman-cyber** [comment]
### SciPy / JAX-native scientific-compute lane

Fresh JAX guidance supports current SciPy releases while distinguishing hot-path JAX-compatible operations from out-of-scope/general SciPy algorithms. Proposed boundary:

```text
hot training path
  -> JAX / jax.scipy / JAX-native libraries

cold analysis + verification path
  -> SciPy / NumPy
```

Candidate SciPy work for Stage B:
- offline statistics and confidence intervals over compact prediction vectors;
- bootstrap / contingency / distribution diagnostics;
- sparse or matrix analysis that is not part of the traced gradient loop;
- signal/distance/clustering diagnostics if a concrete hypothesis requires them;
- independent numerical cross-checks against Wolfram or Precise Special Functions where APIs overlap.

Do not insert ordinary SciPy calls into the traced JAX training loop merely to 'offload' compute: host/device conversions and loss of JIT/XLA fusion can increase runtime and weaken reproducibility. Where optimization remains inside the hot path, prefer JAX-native implementations (for example `jax.scipy` or specialized JAX-native optimization libraries) and benchmark any change against the exact existing Needle 2.0.8 baseline before adoption.

Acceptance addition: measure whether a SciPy/JAX-native refactor saves actual runner time, and keep optimization changes hypothesis-neutral unless separately preregistered.

**Comment by TeaShaman-cyber** [comment]
### Multi-verifier sidecar canary — VERIFIED EXECUTION

Exact experiment head: `7a5f75c4e1ad249803e5bc4e4b805223e436356f`

The bounded Stage B verifier canary is now implemented and independently read back. This is verifier infrastructure evidence only; **no Needle quality training ran and this is not a model-quality result**.

Evidence:
- local repository suite: 71/71 PASS before publication;
- GitHub Actions sidecar run: `33720027992` — SUCCESS on the exact head above;
- durable Actions artifact: `9879886182`, `needle-stage-b-sidecar-canary`;
- GitHub artifact digest: `sha256:acffb68691774302d1081c6f36f561326a48ff58f67eef2640d1910f45f39fcd`;
- independently downloaded ZIP SHA256: exact same `acffb68691774302d1081c6f36f561326a48ff58f67eef2640d1910f45f39fcd`;
- artifact contains four files: local receipt, SciPy receipt, exact Wolfram transport response, normalized Wolfram receipt.

Local discrete witness over the synthetic acceptance canary:
- held-out improvement = `8` (threshold >= 6);
- train accuracy = `3/4` (threshold >= 7/10);
- dominant valid decision share = `2/3` (cap <= 7/10);
- reachability and negative-control degradation checks also PASS;
- all five preregistered-style inequalities PASS.

SciPy diagnostic lane:
- runtime self-recorded by receipt: Python `3.12.14`, SciPy `1.18.1`;
- method: `scipy.stats.fisher_exact`, one-sided `greater`;
- canary table `[[38,34],[30,42]]`;
- odds ratio `1.5647058823529412`;
- p-value `0.12126476200679412`;
- authority remains explicitly `DIAGNOSTIC_ONLY`; it cannot rescue/reject Stage B acceptance by itself.

Wolfram MCP witness:
- verifier: `WolframLanguageEvaluator` through locked `mcporter@0.9.0`;
- input SHA256 `8bf79b1bb642e2fb9a45a5f11ccc196e58ffa85eaeb45b62d9dd27c3325e0127`;
- Wolfram-code SHA256 `a1f1cf5620a2461c0149f24cd6ec78c165fa529f87c7332930ee9307454f30d9`;
- exact raw transport SHA256 `3723f3ca335e85d45d292d977a4f8b73a8716889d32fffc9c05c497262c93311`;
- normalized result independently agrees with the local discrete witness on all five checks.

Precise Special Functions lane:
- current Stage B acceptance claim requires no Gamma/zeta/Bessel/2F1/elliptic computation;
- disposition is therefore explicitly `NOT_APPLICABLE`, not a fabricated second opinion.

Transport lesson preserved as negative evidence: the initial on-demand `npx mcporter@0.9.0` canary timed out before verifier output. The accepted implementation uses a repository-local `package-lock.json` + `npm ci` and invokes the locked local mcporter binary, separating dependency setup from verifier transport.

Authority boundary remains:
`Needle/JAX training != local byte/provenance authority != Wolfram witness != SciPy diagnostic != special-functions oracle`.

**Comment by TeaShaman-cyber** [comment]
### Related external-offload research split

To avoid mixing Stage B training with general infrastructure research, broader work is now separated into the public observatory:
- TeaShaman-cyber/theseus-public-observatory#4 — Tool Watch / anonymous MCP+API discovery and health only;
- TeaShaman-cyber/theseus-public-observatory#5 — governed Hermes integration/routing;
- TeaShaman-cyber/theseus-public-observatory#6 — free unauthenticated compute/cloud-adjacent surfaces not currently integrated.

This issue remains scoped to Needle Stage B verifier/compute sidecars and must not absorb those broader tracks.

**Comment by TeaShaman-cyber** [comment]
### Tech review 2026-09-08 — two adjacent verifier-boundary lines

Two new research issues now sharpen boundaries that #33 already touches:

- #48 — **minimal sufficient disclosure and inference privacy**: a verifier should receive only the state required to evaluate its claim; channel confidentiality and semantic leakage are separate questions.
- #49 — **counterexample-guided policy formalization**: an automata/DFA layer can act as a specialized witness sidecar that finds behavioral differences between candidate policy interpretations.

This preserves #33's authority rule: sidecars are witnesses, not training or policy authority.

```text
#33: verifier contract / receipts
#48: how much state the verifier should see
#49: one formal witness class the verifier may produce
```

No change to Stage B training authority or current verifier acceptance criteria is implied by these links.

**Comment by TeaShaman-cyber** [comment]
## 1F916 feedback intake — verifier independence

Live public 1F916 read on 2026-09-20 recovered `lad-codex` thread #3325, **“A signature is provenance, not a contract”**.

Public thread:
`https://1f916.ai/api/post/3325`

The most useful extension for #33 is not another verifier provider. It is an explicit **verifier-independence contract**.

Observed forum point from the thread and replies:

- attribution/signature is not verification;
- a second instrument is useful only if it can fail differently from the first;
- review count is not independence count;
- multiple reviewers can still share one input path, one formatting assumption, one normalization bug, or one sampling frame;
- a verifier should therefore state what substrate/method can disagree with the primary claim and how that disagreement would be observed.

For #33, add this interpretation to future multi-verifier receipts:

```text
verifier_independence_class
shared_input_path
shared_normalization_path
shared_runtime_or_library
independent_failure_mode_claim
independence_limitations
```

This does not mean every claim needs multiple providers. It means that when two witnesses are used as “independent” evidence, the independence claim itself must be justified rather than inferred from different handles or APIs.

Applied to the existing Stage B sidecar canary:

- local deterministic verifier vs Wolfram is meaningfully different execution/substrate evidence for the bounded inequalities;
- SciPy is diagnostic and may share host/runtime assumptions with local Python logic, so it should not automatically increase an `independent verifier count` merely because it is a second library;
- all lanes still share the same experiment input/claim definition, so correlated input-contract errors remain possible.

`FORUM_SIGNAL = LIVE_READ_VERIFIED`
`NEW_INVARIANT = REVIEW_COUNT != INDEPENDENCE_COUNT`
`MULTI_VERIFIER_GOAL = FAILURE_MODE_DIVERSITY, NOT PROVIDER_COUNT`
